### DISCUSSION TITLE
Ability to rebuild image from scratch

### DISCUSSION THREAD

**Discussion by dozer111** [werf/werf#6362] [General]
### Before proceeding

- [X] I didn't find a similar [issue](https://github.com/werf/werf/issues)

### Problem

Hi

Rarely I need to rebuild the image from the scratch. It is very tricky now, because i need to do some things(e.g change the werf.yaml to push werf to generate new image instead cached one)

Seems cool idea to have a flag like "-no-cache", so when I run "werf build --no-cache" - it would ignore all the cache and build the new image

**Comment by alexey-igrychev** [comment]
@dozer111, Hi, I thank you for the question.

The main goal here is to trigger a rebuild while preserving the determinism of the build and ensuring that previous builds remain reproducible.

At the moment, for such tasks in werf, we recommend explicitly initiating a rebuild in the configuration using the `fromCacheVersion` directive and other directives with the `...CacheVersion` prefix. This allows you to control which parts of the build should be rebuilt while still maintaining reproducibility.

Furthermore, you can configure a deterministic periodic rebuild like this:

```yaml
fromCacheVersion: {{ div .Commit.Date.Unix (mul 60 60 24 30) }} # rebuild every month
```

Currently, this functionality is only available for stapel images, while for Dockerfile-based builds, you can achieve similar behavior using `ARG`.

In upcoming releases, we plan to extend the `fromCacheVersion` feature to Dockerfile builds and provide global control through the `build.fromCacheVersion` directive and a corresponding CLI option.

**Comment by isavcic** [comment]
So, when the code changes, I need to change the Dockerfile as well? And there is no build option flag to override this? You are really, really pushing and forcing the determinism too far while catering only for a very narrow user group and very narrow use case.

**Reply by alexey-igrychev** [reply]
Hi!

This is not about determinism but rather about the content+config-based tagging approach. If the image's content or configuration doesn't change, a rebuild doesn’t happen—we reuse the existing image.

We understand that there’s currently no convenient way to update the configuration version for all images at once, and Dockerfile has certain limitations. We are already addressing these issues and plan to resolve them. Moreover, we might even elevate this functionality to the CLI level for easier usage.
